### PR TITLE
Q8_0: wide int8 matmul, eight output rows per 256-bit load

### LINKED ISSUE #51: CPU Q8_0 matmul with int8 activations and an eight-row tile: 1.13–1.25x prefill, 1.16–1.23x decode — interested?
Hi. I have been comparing SharpMind's CPU path with llama.cpp and with [pulsarforge](https://github.com/siris9476/pulsarforge), a small MIT-licensed C engine that matches llama.cpp on a laptop CPU with AVX2 only. Two of its techniques port cleanly to SharpMind's kernels:
- The attention one is a drop-in kernel change and is up as #50.
- The Q8_0 matmul changes memory, numerics and dispatch. Before opening a PR for it, I would like to know whether you want it and how you would like those points handled.

The code is on my fork, [`3509ddc`](https://github.com/MBrekhof/SharpMind/commit/3509ddc) (branch `perf/pulsarforge-lessons`).

### Q8_0 matmul: eight output rows per 256-bit load, int8 activations

**How it works.**
- **Repack.** Each Q8_0 weight tensor is regrouped eight output rows at a time. Per 32-column block, the eight row scales are stored as float32, followed by eight quartets that hold the eight rows' four bytes side by side. One ymm load therefore holds four columns of eight rows.
- **Activations.** They are quantized to int8 per 32 values, as llama.cpp's CPU q8_0 path does.
- **Dot products.** The activation quartet is broadcast to all eight lanes, and abs/sign + maddubs + madd give the eight rows' dot products, with one sign pair per 32 products.
- **Threads.** They split by pairs of row groups, each sweeping every token of the chunk.

**Scope and tests.**
- It is used only where the parallel FMA Q8_0 kernel is selected and K % 32 == 0. Rows past the last group of eight fall back to `VecDotQ8_0_FMA`.
- The repack is built on first CPU use, so GPU-served layers never build it.
- The tests cover:
  - a double-precision reference of the same quantized arithmetic, computed off the raw blocks;
  - a bound against the float kernel;
  - concurrency and cache single-build;
  - the tiny-activation edge case;
  - kernel selection.
- The tests are mutation-checked.

**Performance.** Measured on one Zen 5 laptop (Ryzen AI 7 350) on AC power, arms interleaved, 3–5 passes. Figures are tok/s for pp512 / tg64 @512 / pp512 @512.

| model | llama.cpp b8683 CPU | SharpMind with | without | gain |
|---|---|---|---|---|
| Qwen3-0.6B Q8_0 | 443 / 49.1 / 384 | 400 / 47.4 / 299 | 344 / 39.3 / 265 | 1.16x / 1.21x / 1.13x |
| Qwen2-0.5B Q8_0 | 609 / 70.0 / 538 | 520 / 63.7 / 473 | 452 / 52.0 / 413 | 1.15x / 1.23x / 1.14x |
| Qwen2.5-1.5B Q8_0 | 172 / 23.9 / 162 | 150 / 21.6 / 139 | 120 / 18.7 / 113 | 1.25x / 1.16x / 1.23x |

The Qwen3 arms use master's attention kernel. The other two models include the #50 kernel in both arms.

**Quality.** Perplexity over the same text, 1091 tokens, teacher-forced. The llama.cpp deltas come from `llama-perplexity` on the same file, using its own protocol.

| model | F16 → Q8_0 float activations → Q8_0 int8 activations | llama.cpp F16 → Q8_0 |
|---|---|---|
| Qwen2-0.5B | 66.25 → 66.32 → 66.47 (+0.33% vs F16) | +0.46% |
| Qwen2.5-1.5B | 40.16 → 40.03 → 40.17 (+0.02% vs F16) | +0.13% |

### Design questions

1. **Memory.** The repacked copy (5.9% larger than the Q8_0 bytes) is kept next to the raw bytes, because the GPU plugin, SMM export and the tail rows read `RawQuantizedData`. That costs about the Q8_0 weight bytes again on CPU: +678 MB on Qwen3-0.6B, +559 MB on Qwen2-0.5B and +1.6 GB on Qwen2.5-1.5B.
   - For scale: with #46 and #47 applied, the post-load footprint of those models is 1720 / 1830 / 4896 MB. The raw-plus-repack cost is a significant share of that.
   - Keep both, or replace the raw bytes with the repacked form and teach the other readers?
2. **Numerics.** CPU Q8_0 projections now carry activation-quantization noise, so CPU and GPU logits on Q8_0 models agree to about 0.1 rather than 1e-3. One GPU parity test would need a wider bound. I would add a check that the wrong head still moves the logits far more than that.
3. **Dispatch.** The change bypasses the JigSaw kernel entry for `q8_0_parallel_fma` from inside `InferenceLinearLayer` / `LogitOps` instead of registering a new kernel, because it needs the repacked data rather than a raw pointer. Is that acceptable, or would you rather see a different seam?
4. **First use.** The repack happens on the first forward: 44–426 ms on these models, which the CUI and server warm-ups already absorb. Keep that, or repack at load?

Happy to shape the PR around your answers.

### PR DESCRIPTION
**Pull request by MBrekhof**
Closes #51.

CPU Q8_0 matmul with int8 activations and an eight-row tile, shaped by your answers on #51:

- **Memory:** the repacked copy is kept next to the raw bytes. The cost is the one quoted in #51. Replacing the raw bytes can be a follow-up PR.
- **Numerics:** CPU Q8_0 projections on this path carry int8 activation-quantization noise (see Quality below). In #51 I said one GPU parity test would need a wider bound. It does not, so no test changes: that test's fixture runs the scalar tier, which never selects this path, and it still passes at 1e-3. No test currently compares FMA-tier CPU logits with the GPU on Q8_0.
- **Dispatch:** `InferenceLinearLayer` and `LogitOps` call the wide kernel directly where `q8_0_parallel_fma` is selected and K % 32 == 0. Rows past the last group of eight use `VecDotQ8_0_FMA`.
- **First use:** the repack stays lazy. It is built under a lock on the first CPU forward through a layer. Layers the GPU serves never build it unless the engine falls back to the CPU.

Technique from [pulsarforge](https://github.com/siris9476/pulsarforge) (MIT). How the repack and the dot products work is in #51 and the commit message.

### Performance

**Setup**
- One Zen 5 laptop (Ryzen AI 7 350, 8C/16T) on AC power.
- Both builds run the same harness; the only difference is the commit: `origin/master` 1f3f1a7 (which already includes #50) against this branch.
- Nine passes. The arms are interleaved inside each pass, and the order swaps every pass.
- Each run loads the model, warms up with pp512 + tg16 and pp1024 + tg16, then times three reps.
  - All figures are warm, so the one-off repack cost quoted in #51 is not in them.
  - Decode feeds a fixed token.
- Prefill runs in chunks of 64, as the app does.

**Reading the table**
- The gain is the median of the nine paired per-pass ratios, and the range is the lowest and highest of those ratios.
- The tok/s columns are each arm's median over the nine passes.
- Throughput drifted during the Qwen3 run: master's pp512 fell 14% from pass 1 to pass 9. Pairing within a pass makes the gain less sensitive to that drift. It is also why the Qwen3 decode gain (1.24x) differs from the quotient of its two columns (1.17x).

| model | test | gain | per-pass range | master (tok/s) | this PR (tok/s) |
|---|---|---|---|---|---|
| Qwen3-0.6B Q8_0 | pp512 | 1.18x | 1.17–1.20x | 364.6 | 425.1 |
| | tg64 @ d512 | 1.24x | 1.15–1.33x | 40.1 | 47.0 |
| | pp512 @ d512 | 1.14x | 1.13–1.20x | 296.7 | 346.6 |
| Qwen2-0.5B Q8_0 | pp512 | 1.16x | 1.14–1.19x | 409.8 | 475.0 |
| | tg64 @ d512 | 1.24x | 1.20–1.27x | 52.0 | 64.7 |
| | pp512 @ d512 | 1.15x | 1.13–1.17x | 372.8 | 427.8 |
| Qwen2.5-1.5B Q8_0 | pp512 | 1.27x | 1.21–1.32x | 119.2 | 150.2 |
| | tg64 @ d512 | 1.14x | 1.12–1.21x | 19.1 | 22.1 |
| | pp512 @ d512 | 1.25x | 1.18–1.29x | 112.1 | 139.6 |

The PR arm was faster in all 81 per-pass comparisons.

<details><summary>Per-pass means (tok/s: pp512, tg64 @ d512, pp512 @ d512)</summary>

Qwen3-0.6B-Q8_0
```
pass 1 master   368.8  39.48   292.8
pass 1 PR       443.8  52.55   346.6
pass 2 PR       443.3  52.85   361.6
pass 2 master   374.6  41.35   317.7
pass 3 master   372.2  40.10   315.2
pass 3 PR       443.3  52.06   366.3
pass 4 PR       440.3  50.36   352.8
pass 4 master   371.1  40.46   312.3
pass 5 master   364.6  39.80   306.5
pass 5 PR       425.1  46.85   349.8
pass 6 PR       421.3  46.86   346.4
pass 6 master   351.0  40.68   296.7
pass 7 master   344.7  40.24   290.4
pass 7 PR       405.0  46.95   330.0
pass 8 PR       382.0  46.59   321.1
pass 8 master   325.9  37.05   268.6
pass 9 master   318.2  38.85   267.8
pass 9 PR       376.4  46.83   304.9
```

qwen2-0_5b-instruct-q8_0
```
pass 1 master   401.9  53.63   372.7
pass 1 PR       479.3  65.23   429.7
pass 2 PR       474.8  64.68   427.1
pass 2 master   413.7  51.98   377.0
pass 3 master   407.8  53.67   367.0
pass 3 PR       475.0  65.05   429.0
pass 4 PR       471.4  64.60   424.5
pass 4 master   409.7  53.33   376.9
pass 5 master   409.8  51.82   370.1
pass 5 PR       473.8  64.22   426.5
pass 6 PR       480.8  65.26   431.3
pass 6 master   408.2  51.74   372.8
pass 7 master   415.9  53.88   378.8
pass 7 PR       475.7  64.60   427.3
pass 8 PR       469.0  64.67   427.8
pass 8 master   410.8  51.86   371.1
pass 9 master   410.4  51.29   373.7
pass 9 PR       475.4  64.92   428.3
```

qwen2.5-1.5b-instruct-q8_0
```
pass 1 master   117.3  18.72   111.6
pass 1 PR       141.6  21.38   131.6
pass 2 PR       147.6  21.83   139.2
pass 2 master   119.2  19.14   112.1
pass 3 master   118.3  19.00   111.6
pass 3 PR       150.0  22.18   139.5
pass 4 PR       150.0  22.16   139.5
pass 4 master   114.2  18.28   108.4
pass 5 master   113.8  19.12   109.0
pass 5 PR       150.2  22.14   139.6
pass 6 PR       151.1  21.91   140.5
pass 6 master   120.6  19.60   113.4
pass 7 master   120.1  19.46   112.1
pass 7 PR       152.4  22.19   141.6
pass 8 PR       153.0  22.02   143.9
pass 8 master   122.1  19.49   115.1
pass 9 master   124.0  19.59   116.6
pass 9 PR       158.5  22.21   147.1
```

</details>

### Quality

I re-checked perplexity on this base, on the same text as in #51: Qwen2-0.5B, 1095 teacher-forced tokens through the decode path.

| arm | perplexity | change |
|---|---|---|
| F16 | 66.37 | — |
| Q8_0, float activations | 66.44 | +0.10% vs F16 |
| Q8_0, this PR | 66.59 | +0.24% vs float activations, +0.34% vs F16 |

On that text, llama.cpp's own F16 → Q8_0 change is +0.46%. Top-1 is unchanged. Positions 6–10 of the top-10, where the logits are nearly tied, reorder within about 0.1 logit.

### Tests

`Q8_0WideWeightsTests` covers:
- a double-precision reference of the same quantized arithmetic, computed from the raw blocks;
- a bound against the float kernel;
- concurrency: parallel runs reproduce serial results, 64 concurrent first calls share one repack, and a new raw array gets a new one;
- the tiny-activation edge case;
- kernel selection: only a parallel FMA selection enables the path.

Full suite: 1674 passed, 0 failed.

### CODE REVIEW

**Review by Integral2u — APPROVED** [review]
Once again mammoth effort.  thank you.
